Fix memory leak issue in `linearize`

agrawalraj opened the pull request:
This PR adds an argument `detach` to `linearize` and `make_empirical_fisher_vp` to avoid an odd memory leak when gradients are not detached in `make_empirical_fisher_vp`. In case people would like to autodiff through `linearize` (e.g., for higher-order influence function computations), then `detach` should be set to `False`. 

Resolves #516.

SamWitty commented:
@agrawalraj , I'm not seeing a test that confirms that this addresses the memory leak issue. Could you please add a test that shows: (i) when using `detach=False` the memory footprint increases for two successive calls to `influence_fn` and (ii) when using `detach=True` the memory footprint does not increase for two successive calls?

SamWitty commented on chirho/robust/ops.py:
```diff
@@ -127,7 +126,7 @@ def _influence_functional(*models: Callable[P, Any]) -> Callable[P, S]:
             raise ValueError("mismatch between number of models and points")
 
-        linearized = linearize(*models, **linearize_kwargs)
+        linearized = linearize(*models, **linearize_kwargs)  # type: ignore
```
Why is this needed?

agrawalraj replied:
This is unrelated / can remove this line. I added it because the linter complains that linearized isn't typed.

agrawalraj commented:
@SamWitty I added `test_memory_leak.py` to track how memory changes when `detach` is `True` or `False`. I didn't make it a formal test because this issue might be specific to `torch.func` and could be resolved in a future release of `torch.func`. The memory leak when `detach = False` is weird because I'm not saving any variables in the loop...

SamWitty commented:
> @SamWitty I added `test_memory_leak.py` to track how memory changes when `detach` is `True` or `False`. I didn't make it a formal test because this issue might be specific to `torch.func` and could be resolved in a future release of `torch.func`. The memory leak when `detach = False` is weird because I'm not saving any variables in the loop...

Fair enough. I think a good approach then is to just add a version with `detach = True` to the tests, making sure that it doesn't accumulate a memory footprint. This way we'll be able to confirm that this actually addresses the memory bug, while having the tests be robust to upstream changes in `torch.func`.

eb8680 commented:
Can this be addressed using the [`torch.no_grad()`](https://pytorch.org/docs/stable/generated/torch.no_grad.html#no-grad) context manager without adding a flag to the API?

agrawalraj commented:
> Can this be addressed using the [`torch.no_grad()`](https://pytorch.org/docs/stable/generated/torch.no_grad.html#no-grad) context manager without adding a flag to the API?

I tried adding this flag and that works! Should we just close this PR then? My only concern though would be people might not know to use `torch.no_grad()` and then might run into this memory issue.

agrawalraj commented:
> > @SamWitty I added `test_memory_leak.py` to track how memory changes when `detach` is `True` or `False`. I didn't make it a formal test because this issue might be specific to `torch.func` and could be resolved in a future release of `torch.func`. The memory leak when `detach = False` is weird because I'm not saving any variables in the loop...
> 
> Fair enough. I think a good approach then is to just add a version with `detach = True` to the tests, making sure that it doesn't accumulate a memory footprint. This way we'll be able to confirm that this actually addresses the memory bug, while having the tests be robust to upstream changes in `torch.func`.

Good idea. I changed the file to explicitly test with `detach=True`. However, we might abandon this PR b/c you can avoid this memory issue w/ `torch.no_grad()`

agrawalraj commented:
Closing this PR in favor of #519.
